cub/agent/agent_batch_memcpy.cuh

```diff
@@ -0,0 +1,967 @@
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+/**
+ * \file
+ * cub::AgentBatchMemcpy implements device-wide copying of a batch of device-accessible source-buffers to
+ * device-accessible destination-buffers.
+ */
+
+#pragma once
+
+#include <cub/agent/single_pass_scan_operators.cuh>
+#include <cub/block/block_exchange.cuh>
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_run_length_decode.cuh>
+#include <cub/block/block_scan.cuh>
+#include <cub/block/block_store.cuh>
+#include <cub/util_ptx.cuh>
+#include <cub/util_type.cuh>
+
+#include <cstdint>
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+__forceinline__ __device__ void LoadVectorAndFunnelShiftR(uint32_t const *aligned_ptr,
+                                                          uint32_t bit_shift,
+                                                          uint4 &data_out)
+{
+  data_out      = {aligned_ptr[0], aligned_ptr[1], aligned_ptr[2], aligned_ptr[3]};
+  uint32_t tail = 0;
+  if (bit_shift)
+    tail = aligned_ptr[4];
+
+  data_out.x = __funnelshift_r(data_out.x, data_out.y, bit_shift);
+  data_out.y = __funnelshift_r(data_out.y, data_out.z, bit_shift);
+  data_out.z = __funnelshift_r(data_out.z, data_out.w, bit_shift);
+  data_out.w = __funnelshift_r(data_out.w, tail, bit_shift);
+}
+
+__forceinline__ __device__ void LoadVectorAndFunnelShiftR(uint32_t const *aligned_ptr,
+                                                          uint32_t bit_shift,
+                                                          uint2 &data_out)
+{
+  data_out      = {aligned_ptr[0], aligned_ptr[1]};
+  uint32_t tail = 0;
+  if (bit_shift)
+    tail = aligned_ptr[2];
+
+  data_out.x = __funnelshift_r(data_out.x, data_out.y, bit_shift);
+  data_out.y = __funnelshift_r(data_out.y, tail, bit_shift);
+}
+
+__forceinline__ __device__ void LoadVectorAndFunnelShiftR(uint32_t const *aligned_ptr,
+                                                          uint32_t bit_shift,
+                                                          uint32_t &data_out)
+{
+  data_out      = aligned_ptr[0];
+  uint32_t tail = 0;
+  if (bit_shift)
+    tail = aligned_ptr[1];
+
+  data_out = __funnelshift_r(data_out, tail, bit_shift);
+}
+
+template <typename VectorT>
+__forceinline__ __device__ void LoadVector(const char *ptr, VectorT &data_out)
+{
+  const uint32_t offset            = reinterpret_cast<std::uintptr_t>(ptr) % 4U;
+  const uint32_t *aligned_ptr      = reinterpret_cast<uint32_t const *>(ptr - offset);
```

Comment:
A comment on why it's safe to always load 32 bits would help a great deal :)

Reply:
I've added a comment that outlines the behaviour of this function.